src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/LanguageServices/Handlers/MetadataReferenceItemHandler.cs

```diff
@@ -54,25 +55,72 @@ public void Handle(IComparable version, BuildOptions added, BuildOptions removed
-                _context.AddMetadataReference(fullPath, properties);
-                bool added = _paths.Add(fullPath);
-                Assumes.True(added);
+                // The reference has already been added previously. The current implementation of IWorkspaceProjectContext
+                // presumes that we'll only called AddMetadataReference once for a given path. Thus we have to remove the
+                // existing one, compute merged properties, and add the new one.
+                _context.RemoveMetadataReference(fullPath);
+
+                var combinedAliases = GetEmptyIfGlobalAlias(GetGlobalAliasIfEmpty(existingProperties.Aliases).AddRange(GetGlobalAliasIfEmpty(properties.Aliases)));
+                properties = properties.WithAliases(combinedAliases);
             }
+
+            logger.WriteLine("Adding reference '{0}'", fullPath);
+            _context.AddMetadataReference(fullPath, properties);
+            _addedPathsWithMetadata[fullPath] = properties;
         }
 
-        private void RemoveFromContextIfPresent(string fullPath, IProjectLogger logger)
+        private void RemoveFromContextIfPresent(string fullPath, MetadataReferenceProperties properties, IProjectLogger logger)
         {
-            if (_paths.Contains(fullPath))
+            if (_addedPathsWithMetadata.TryGetValue(fullPath, out var existingProperties))
             {
-                logger.WriteLine("Removing reference '{0}'", fullPath);
                 _context.RemoveMetadataReference(fullPath);
-                bool removed = _paths.Remove(fullPath);
-                Assumes.True(removed);
+
+                // Subtract any existing aliases out. This will be an empty list if we should remove the reference entirely
+                var resultantAliases = GetGlobalAliasIfEmpty(existingProperties.Aliases).RemoveRange(GetGlobalAliasIfEmpty(properties.Aliases));
+
+                if (resultantAliases.IsEmpty)
+                {
+                    // There's nothing left here, completely remove it
+                    Assumes.True(_addedPathsWithMetadata.Remove(fullPath));
+                }
+                else
+                {
+                    // resultantAliases might be the global alias. In that case, let's remove it again.
+                    resultantAliases = GetEmptyIfGlobalAlias(resultantAliases);
+                    properties = properties.WithAliases(resultantAliases);
+                    _context.AddMetadataReference(fullPath, properties);
+                    _addedPathsWithMetadata[fullPath] = properties;
+                }
             }
         }
+
+        /// <summary>
+        /// Returns the list of aliases, replacing an empty list with "global".
+        /// </summary>
+        private static ImmutableArray<string> GetGlobalAliasIfEmpty(ImmutableArray<string> aliases)
+        {
+            if (aliases.IsDefaultOrEmpty)
+            {
+                return ImmutableArray.Create("global");
```

Comment:
This is going to be a common case. Should we extract this out into a field so the object can be reused, rather than creating a new one each time?